File: src/main/scala/com/yj/scala/transformation/TestJoin.scala
package com.yj.scala.transformation

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
 * 父RDD的分区数与子RDD中分区数较多的那个保持一致
 *
 */
/**
 * leftOuterJoin
 * rightOuterJoin
 * fullOuterJoin
 */
case object TestJoin {
  def main(args: Array[String]): Unit = {

    val context: SparkContext = new SparkContext(new SparkConf()
      .setMaster("local")
      .setAppName("Join test"))
    val ageRDD: RDD[(String, Int)] = context.parallelize(Array[(String, Int)](
      ("张三", 18),
      ("李四", 19),
      ("王5", 20),
      ("猪8戒", 500)
    ))
    val scoreRDD: RDD[(String, Int)] = context.parallelize(Array[(String, Int)](
      ("张三", 300),
      ("李四", 500),
      ("王5", 800),
      ("马六", 700)
    ))
    val join: RDD[(String, (Int, Option[Int]))] = ageRDD.leftOuterJoin(scoreRDD)
    join.foreach(println)
    context.stop()
  }
}

File: src/main/java/com/yj/it/transformation/TestJoin.java
package com.yj.it.transformation;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

import java.util.Arrays;

/**
 * join : 作用在K,V格式的RDD上，（K,V）.join(K,W) => (K,(V,W)) 两个RDD相同的key才能关联在一起
 */
public class TestJoin {
    public static void main(String[] args) {
        JavaSparkContext context = new JavaSparkContext(new SparkConf()
                .setMaster("local")
                .setAppName("join test"));

        JavaPairRDD<String, Integer> ageRDD = context.parallelizePairs(Arrays.asList(
                new Tuple2<>("张三", 18),
                new Tuple2<>("李四", 25),
                new Tuple2<>("王5", 19)
        ));
        JavaPairRDD<String, Integer> scoreRDD = context.parallelizePairs(Arrays.asList(
                new Tuple2<>("张三", 89),
                new Tuple2<>("李四", 100),
                new Tuple2<>("王5", 55)
        ));

        JavaPairRDD<String, Tuple2<Integer, Integer>> join = ageRDD.join(scoreRDD);
        join.foreach(new VoidFunction<Tuple2<String, Tuple2<Integer, Integer>>>() {
            @Override
            public void call(Tuple2<String, Tuple2<Integer, Integer>> stringTuple2Tuple2) throws Exception {
                System.out.println(stringTuple2Tuple2);
            }
        });

        context.stop();
    }
}
